engine/evd.ml

```diff
@@ -205,21 +205,26 @@ module Store = Store.Make ()
 
 let string_of_existential evk = "?X" ^ string_of_int (Evar.repr evk)
 
-type evar_body =
-  | Evar_empty
-  | Evar_defined of constr
+type defined = [ `defined ]
+type undefined = [ `undefined ]
 
-type evar_info = {
+type _ evar_body =
+  | Evar_empty : undefined evar_body
+  | Evar_defined : econstr -> defined evar_body
+
+type 'a evar_info = {
   evar_concl : constr;
   evar_hyps : named_context_val;
-  evar_body : evar_body;
+  evar_body : 'a evar_body;
   evar_filter : Filter.t;
   evar_abstract_arguments : Abstraction.t;
   evar_source : Evar_kinds.t Loc.located;
   evar_candidates : constr list option; (* if not None, list of allowed instances *)
   evar_relevance: Sorts.relevance;
 }
 
+type any_evar_info = EvarInfo : 'a evar_info -> any_evar_info
```

Comment:
I suggest to add `[@ocaml.unboxed]`, since the constructor is only there for typing reasons, not for operational reasons, so as to reduce the stress on the garbage collector and avoid tons of indirections.

Reply:
I don't think it matters because this is only returned by `find`, so we never get indirections in evarmaps for instance. That said it cannot hurt, so I will tweak it anyways.

Reply:
Ah, the type-checker is not happy.
```
Error: This type cannot be unboxed because
       it might contain both float and non-float values,
       depending on the instantiation of the existential variable 'a.
       You should annotate it with [@@ocaml.boxed].
```

Reply:
Damn. As for the indirections, I was talking about all the statements `let EvarInfo evi =`. They are indirections and they do have a cost.

Reply:
I really doubt that this kind of indirection is problematic in OCaml. These are extremely short-lived so that they are always kept in the minor heap, which is basically free with our GC settings. I also think flambda optimizes them out even without the unboxed flag. Finally, it's not so clear we perform that many Evd.find, from my experience most of the calls are Evd.find_undefined which doesn't box.

Reply:
Even if the allocations are short-lived, they still fill the minor heap, which in turn causes a garbage collection once full. (For OCaml 5, the situation is even worse, since collecting the minor heap is a stop-the-world garbage collection. If we ever have a multicore Coq, as soon as a single thread fills its minor heap, all the other threads will come to a halt.) As for Flamba, it only optimizes these allocations if the functions are inlined and if there is no exception handler. Anyway, if you are confident that `Evd.find` is hardly ever called, that is fine with me.